pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinerunresolution_test.go

```diff
@@ -561,3 +561,283 @@ func TestGetPipelineConditionStatus(t *testing.T) {
 		})
 	}
 }
+
+func TestValidateProvidedBy(t *testing.T) {
```

Comment:
You could use the `test/builder` package here :angel: :wink:

Reply:
Looks like I can definitely use the resource builders but I'm not sure I can use the builders to construct the `state` object - I'll explain more in my follow up PR!